{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f6d1eb0-6fba-423d-ae16-495b21680a40",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q1. Explain why we have to use the Exception class while creating a Custom Exception."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a469dee-d027-4b35-9193-2fc17f84cd79",
   "metadata": {},
   "outputs": [],
   "source": [
    "In object-oriented programming, an exception is an event or condition that disrupts the normal flow of a program's execution. \n",
    "Exceptions are used to handle errors, unexpected situations, or other exceptional conditions that may occur during the runtime of a program. \n",
    "Custom exceptions are exceptions that you define yourself, tailored to specific scenarios in your codebase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e80e68a-4afa-43af-a94b-95e72c73bf3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e4c3e0b-115e-433e-8a3b-8568c9852913",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'BaseException'>\n",
      "  <class 'Exception'>\n",
      "    <class 'TypeError'>\n",
      "      <class 'decimal.FloatOperation'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "    <class 'StopAsyncIteration'>\n",
      "    <class 'StopIteration'>\n",
      "    <class 'ImportError'>\n",
      "      <class 'ModuleNotFoundError'>\n",
      "      <class 'zipimport.ZipImportError'>\n",
      "    <class 'OSError'>\n",
      "      <class 'ConnectionError'>\n",
      "        <class 'BrokenPipeError'>\n",
      "        <class 'ConnectionAbortedError'>\n",
      "        <class 'ConnectionRefusedError'>\n",
      "        <class 'ConnectionResetError'>\n",
      "          <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'BlockingIOError'>\n",
      "      <class 'ChildProcessError'>\n",
      "      <class 'FileExistsError'>\n",
      "      <class 'FileNotFoundError'>\n",
      "      <class 'IsADirectoryError'>\n",
      "      <class 'NotADirectoryError'>\n",
      "      <class 'InterruptedError'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'PermissionError'>\n",
      "      <class 'ProcessLookupError'>\n",
      "      <class 'TimeoutError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'signal.itimer_error'>\n",
      "      <class 'socket.herror'>\n",
      "      <class 'socket.gaierror'>\n",
      "      <class 'ssl.SSLError'>\n",
      "        <class 'ssl.SSLCertVerificationError'>\n",
      "        <class 'ssl.SSLZeroReturnError'>\n",
      "        <class 'ssl.SSLWantWriteError'>\n",
      "        <class 'ssl.SSLWantReadError'>\n",
      "        <class 'ssl.SSLSyscallError'>\n",
      "        <class 'ssl.SSLEOFError'>\n",
      "      <class 'shutil.Error'>\n",
      "        <class 'shutil.SameFileError'>\n",
      "      <class 'shutil.SpecialFileError'>\n",
      "      <class 'shutil.ExecError'>\n",
      "      <class 'shutil.ReadError'>\n",
      "      <class 'urllib.error.URLError'>\n",
      "        <class 'urllib.error.HTTPError'>\n",
      "        <class 'urllib.error.ContentTooShortError'>\n",
      "      <class 'gzip.BadGzipFile'>\n",
      "    <class 'EOFError'>\n",
      "      <class 'asyncio.exceptions.IncompleteReadError'>\n",
      "    <class 'RuntimeError'>\n",
      "      <class 'RecursionError'>\n",
      "      <class 'NotImplementedError'>\n",
      "        <class 'zmq.error.ZMQVersionError'>\n",
      "        <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "      <class '_frozen_importlib._DeadlockError'>\n",
      "      <class 'threading.BrokenBarrierError'>\n",
      "      <class 'concurrent.futures._base.BrokenExecutor'>\n",
      "        <class 'concurrent.futures.thread.BrokenThreadPool'>\n",
      "      <class 'asyncio.exceptions.SendfileNotAvailableError'>\n",
      "      <class 'pkg_resources.ExtractionError'>\n",
      "      <class '_pydevd_bundle.pydevd_vars.VariableError'>\n",
      "    <class 'NameError'>\n",
      "      <class 'UnboundLocalError'>\n",
      "    <class 'AttributeError'>\n",
      "      <class 'dataclasses.FrozenInstanceError'>\n",
      "    <class 'SyntaxError'>\n",
      "      <class 'IndentationError'>\n",
      "        <class 'TabError'>\n",
      "    <class 'LookupError'>\n",
      "      <class 'IndexError'>\n",
      "      <class 'KeyError'>\n",
      "        <class 'jupyter_client.kernelspec.NoSuchKernel'>\n",
      "        <class 'IPython.terminal.pt_inputhooks.UnknownBackend'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ValueError'>\n",
      "      <class 'UnicodeError'>\n",
      "        <class 'UnicodeEncodeError'>\n",
      "        <class 'UnicodeDecodeError'>\n",
      "        <class 'UnicodeTranslateError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'json.decoder.JSONDecodeError'>\n",
      "      <class 'ssl.SSLCertVerificationError'>\n",
      "      <class 'binascii.Error'>\n",
      "      <class '_hashlib.UnsupportedDigestmodError'>\n",
      "      <class 'calendar.IllegalMonthError'>\n",
      "      <class 'calendar.IllegalWeekdayError'>\n",
      "      <class 'dateutil.parser._parser.ParserError'>\n",
      "      <class 'pygments.util.ClassNotFound'>\n",
      "      <class 'IPython.lib.clipboard.ClipboardEmpty'>\n",
      "      <class 'email.errors.MessageDefect'>\n",
      "        <class 'email.errors.NoBoundaryInMultipartDefect'>\n",
      "        <class 'email.errors.StartBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.CloseBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.FirstHeaderLineIsContinuationDefect'>\n",
      "        <class 'email.errors.MisplacedEnvelopeHeaderDefect'>\n",
      "        <class 'email.errors.MissingHeaderBodySeparatorDefect'>\n",
      "        <class 'email.errors.MultipartInvariantViolationDefect'>\n",
      "        <class 'email.errors.InvalidMultipartContentTransferEncodingDefect'>\n",
      "        <class 'email.errors.UndecodableBytesDefect'>\n",
      "        <class 'email.errors.InvalidBase64PaddingDefect'>\n",
      "        <class 'email.errors.InvalidBase64CharactersDefect'>\n",
      "        <class 'email.errors.InvalidBase64LengthDefect'>\n",
      "        <class 'email.errors.HeaderDefect'>\n",
      "          <class 'email.errors.InvalidHeaderDefect'>\n",
      "          <class 'email.errors.HeaderMissingRequiredValue'>\n",
      "          <class 'email.errors.NonPrintableDefect'>\n",
      "          <class 'email.errors.ObsoleteHeaderDefect'>\n",
      "          <class 'email.errors.NonASCIILocalPartDefect'>\n",
      "          <class 'email.errors.InvalidDateDefect'>\n",
      "      <class 'IPython.core.magics.code.MacroToEdit'>\n",
      "      <class 'plistlib.InvalidFileException'>\n",
      "      <class 'pkg_resources._vendor.more_itertools.more.UnequalIterablesError'>\n",
      "      <class 'pkg_resources.extern.packaging.version.InvalidVersion'>\n",
      "      <class 'pkg_resources._vendor.packaging._manylinux._ELFFileHeader._InvalidELFFileHeader'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidWheelFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidSdistFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.specifiers.InvalidSpecifier'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.InvalidMarker'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedComparison'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedEnvironmentName'>\n",
      "      <class 'pkg_resources.extern.packaging.requirements.InvalidRequirement'>\n",
      "        <class 'pkg_resources.RequirementParseError'>\n",
      "      <class 'packaging.version.InvalidVersion'>\n",
      "    <class 'AssertionError'>\n",
      "    <class 'ArithmeticError'>\n",
      "      <class 'FloatingPointError'>\n",
      "      <class 'OverflowError'>\n",
      "      <class 'ZeroDivisionError'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.DivisionUndefined'>\n",
      "      <class 'decimal.DecimalException'>\n",
      "        <class 'decimal.Clamped'>\n",
      "        <class 'decimal.Rounded'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Inexact'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Subnormal'>\n",
      "          <class 'decimal.Underflow'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.FloatOperation'>\n",
      "        <class 'decimal.InvalidOperation'>\n",
      "          <class 'decimal.ConversionSyntax'>\n",
      "          <class 'decimal.DivisionImpossible'>\n",
      "          <class 'decimal.DivisionUndefined'>\n",
      "          <class 'decimal.InvalidContext'>\n",
      "    <class 'SystemError'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ReferenceError'>\n",
      "    <class 'MemoryError'>\n",
      "    <class 'BufferError'>\n",
      "    <class 'Warning'>\n",
      "      <class 'UserWarning'>\n",
      "        <class 'getpass.GetPassWarning'>\n",
      "        <class 'IPython.core.formatters.FormatterWarning'>\n",
      "      <class 'EncodingWarning'>\n",
      "      <class 'DeprecationWarning'>\n",
      "        <class 'IPython.core.interactiveshell.ProvisionalWarning'>\n",
      "      <class 'PendingDeprecationWarning'>\n",
      "      <class 'SyntaxWarning'>\n",
      "      <class 'RuntimeWarning'>\n",
      "        <class 'zmq.asyncio.ProactorSelectorThreadWarning'>\n",
      "        <class 'dateutil.parser._parser.UnknownTimezoneWarning'>\n",
      "        <class 'pkg_resources.PEP440Warning'>\n",
      "      <class 'FutureWarning'>\n",
      "        <class 'IPython.core.completer.ProvisionalCompleterWarning'>\n",
      "      <class 'ImportWarning'>\n",
      "      <class 'UnicodeWarning'>\n",
      "      <class 'BytesWarning'>\n",
      "      <class 'ResourceWarning'>\n",
      "      <class 'dateutil.tz.DeprecatedTzFormatWarning'>\n",
      "      <class 'pkg_resources.PkgResourcesDeprecationWarning'>\n",
      "    <class 'warnings._OptionError'>\n",
      "    <class 'runpy._Error'>\n",
      "    <class 're.error'>\n",
      "    <class 'sre_parse.Verbose'>\n",
      "    <class 'locale.Error'>\n",
      "    <class 'subprocess.SubprocessError'>\n",
      "      <class 'subprocess.CalledProcessError'>\n",
      "      <class 'subprocess.TimeoutExpired'>\n",
      "    <class 'tokenize.TokenError'>\n",
      "    <class 'tokenize.StopTokenizing'>\n",
      "    <class 'inspect.ClassFoundException'>\n",
      "    <class 'inspect.EndOfBlock'>\n",
      "    <class 'traitlets.traitlets.TraitError'>\n",
      "    <class 'copy.Error'>\n",
      "    <class 'concurrent.futures._base.Error'>\n",
      "      <class 'concurrent.futures._base.CancelledError'>\n",
      "      <class 'concurrent.futures._base.TimeoutError'>\n",
      "      <class 'concurrent.futures._base.InvalidStateError'>\n",
      "    <class 'socket._GiveupOnSendfile'>\n",
      "    <class 'struct.error'>\n",
      "    <class 'binascii.Incomplete'>\n",
      "    <class 'asyncio.exceptions.TimeoutError'>\n",
      "    <class 'asyncio.exceptions.InvalidStateError'>\n",
      "    <class 'asyncio.exceptions.LimitOverrunError'>\n",
      "    <class 'asyncio.queues.QueueEmpty'>\n",
      "    <class 'asyncio.queues.QueueFull'>\n",
      "    <class '_queue.Empty'>\n",
      "    <class 'queue.Full'>\n",
      "    <class 'ctypes.ArgumentError'>\n",
      "    <class 'zmq.error.ZMQBaseError'>\n",
      "      <class 'zmq.error.ZMQError'>\n",
      "        <class 'zmq.error.ContextTerminated'>\n",
      "        <class 'zmq.error.Again'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'zmq.error.ZMQBindError'>\n",
      "      <class 'zmq.error.NotDone'>\n",
      "    <class '_pickle.PickleError'>\n",
      "      <class '_pickle.PicklingError'>\n",
      "      <class '_pickle.UnpicklingError'>\n",
      "    <class 'pickle._Stop'>\n",
      "    <class 'argparse.ArgumentError'>\n",
      "    <class 'argparse.ArgumentTypeError'>\n",
      "    <class 'traitlets.config.loader.ConfigError'>\n",
      "      <class 'traitlets.config.loader.ConfigLoaderError'>\n",
      "        <class 'traitlets.config.loader.ArgumentError'>\n",
      "      <class 'traitlets.config.loader.ConfigFileNotFound'>\n",
      "    <class 'traitlets.config.configurable.ConfigurableError'>\n",
      "      <class 'traitlets.config.configurable.MultipleInstanceError'>\n",
      "    <class 'traitlets.config.application.ApplicationError'>\n",
      "    <class 'zlib.error'>\n",
      "    <class 'tornado.util.TimeoutError'>\n",
      "    <class '_curses.error'>\n",
      "    <class 'tornado.concurrent.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.KeyReuseError'>\n",
      "    <class 'tornado.gen.UnknownKeyError'>\n",
      "    <class 'tornado.gen.LeakedCallbackError'>\n",
      "    <class 'tornado.gen.BadYieldError'>\n",
      "    <class 'tornado.gen.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.Return'>\n",
      "    <class 'jupyter_client.channels.InvalidPortNumber'>\n",
      "    <class 'termios.error'>\n",
      "    <class '_lzma.LZMAError'>\n",
      "    <class 'shutil.RegistryError'>\n",
      "    <class 'shutil._GiveupOnFastCopy'>\n",
      "    <class 'configparser.Error'>\n",
      "      <class 'configparser.NoSectionError'>\n",
      "      <class 'configparser.DuplicateSectionError'>\n",
      "      <class 'configparser.DuplicateOptionError'>\n",
      "      <class 'configparser.NoOptionError'>\n",
      "      <class 'configparser.InterpolationError'>\n",
      "        <class 'configparser.InterpolationMissingOptionError'>\n",
      "        <class 'configparser.InterpolationSyntaxError'>\n",
      "        <class 'configparser.InterpolationDepthError'>\n",
      "      <class 'configparser.ParsingError'>\n",
      "        <class 'configparser.MissingSectionHeaderError'>\n",
      "    <class 'jupyter_client.localinterfaces.NoIPAddresses'>\n",
      "    <class 'zipfile.BadZipFile'>\n",
      "    <class 'zipfile.LargeZipFile'>\n",
      "    <class 'entrypoints.BadEntryPoint'>\n",
      "    <class 'entrypoints.NoSuchEntryPoint'>\n",
      "    <class 'jupyter_client.multikernelmanager.DuplicateKernelError'>\n",
      "    <class 'pydoc.ErrorDuringImport'>\n",
      "    <class 'executing.executing.NotOneValueFound'>\n",
      "    <class 'pure_eval.utils.CannotEval'>\n",
      "    <class 'pygments.util.OptionError'>\n",
      "    <class 'bdb.BdbQuit'>\n",
      "    <class 'pdb.Restart'>\n",
      "    <class 'pexpect.exceptions.ExceptionPexpect'>\n",
      "      <class 'pexpect.exceptions.EOF'>\n",
      "      <class 'pexpect.exceptions.TIMEOUT'>\n",
      "    <class 'ptyprocess.util.PtyProcessError'>\n",
      "    <class 'IPython.utils.process.FindCmdError'>\n",
      "    <class 'IPython.utils.path.HomeDirError'>\n",
      "    <class 'IPython.core.profiledir.ProfileDirError'>\n",
      "    <class 'IPython.core.error.IPythonCoreError'>\n",
      "      <class 'IPython.core.error.TryNext'>\n",
      "      <class 'IPython.core.error.UsageError'>\n",
      "      <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "    <class 'IPython.core.error.InputRejected'>\n",
      "    <class 'getopt.GetoptError'>\n",
      "    <class 'pygments.filters.ErrorToken'>\n",
      "    <class 'IPython.core.prefilter.PrefilterError'>\n",
      "    <class 'IPython.core.alias.AliasError'>\n",
      "      <class 'IPython.core.alias.InvalidAliasError'>\n",
      "    <class 'sqlite3.Error'>\n",
      "      <class 'sqlite3.InterfaceError'>\n",
      "      <class 'sqlite3.DatabaseError'>\n",
      "        <class 'sqlite3.InternalError'>\n",
      "        <class 'sqlite3.OperationalError'>\n",
      "        <class 'sqlite3.ProgrammingError'>\n",
      "        <class 'sqlite3.IntegrityError'>\n",
      "        <class 'sqlite3.DataError'>\n",
      "        <class 'sqlite3.NotSupportedError'>\n",
      "    <class 'sqlite3.Warning'>\n",
      "    <class 'IPython.core.interactiveshell.SpaceInInput'>\n",
      "    <class 'xml.dom.DOMException'>\n",
      "      <class 'xml.dom.IndexSizeErr'>\n",
      "      <class 'xml.dom.DomstringSizeErr'>\n",
      "      <class 'xml.dom.HierarchyRequestErr'>\n",
      "      <class 'xml.dom.WrongDocumentErr'>\n",
      "      <class 'xml.dom.InvalidCharacterErr'>\n",
      "      <class 'xml.dom.NoDataAllowedErr'>\n",
      "      <class 'xml.dom.NoModificationAllowedErr'>\n",
      "      <class 'xml.dom.NotFoundErr'>\n",
      "      <class 'xml.dom.NotSupportedErr'>\n",
      "      <class 'xml.dom.InuseAttributeErr'>\n",
      "      <class 'xml.dom.InvalidStateErr'>\n",
      "      <class 'xml.dom.SyntaxErr'>\n",
      "      <class 'xml.dom.InvalidModificationErr'>\n",
      "      <class 'xml.dom.NamespaceErr'>\n",
      "      <class 'xml.dom.InvalidAccessErr'>\n",
      "      <class 'xml.dom.ValidationErr'>\n",
      "    <class 'prompt_toolkit.validation.ValidationError'>\n",
      "    <class 'prompt_toolkit.buffer.EditReadOnlyBuffer'>\n",
      "    <class 'prompt_toolkit.buffer._Retry'>\n",
      "    <class 'prompt_toolkit.layout.layout.InvalidLayoutError'>\n",
      "    <class 'prompt_toolkit.renderer.HeightIsUnknownError'>\n",
      "    <class 'parso.parser.ParserSyntaxError'>\n",
      "    <class 'parso.parser.InternalParseError'>\n",
      "    <class 'parso.python.diff._PositionUpdatingFinished'>\n",
      "    <class 'jedi.inference.helpers.SimpleGetItemNotFound'>\n",
      "    <class 'jedi.inference.utils.UncaughtAttributeError'>\n",
      "    <class 'jedi.inference.base_value.HasNoContext'>\n",
      "    <class 'jedi.inference.arguments.ParamIssue'>\n",
      "    <class 'jedi.api.exceptions._JediError'>\n",
      "      <class 'jedi.api.exceptions.InternalError'>\n",
      "      <class 'jedi.api.exceptions.WrongVersion'>\n",
      "      <class 'jedi.api.exceptions.RefactoringError'>\n",
      "    <class 'jedi.api.helpers.OnErrorLeaf'>\n",
      "    <class 'jedi.api.environment.InvalidPythonEnvironment'>\n",
      "    <class 'email.errors.MessageError'>\n",
      "      <class 'email.errors.MessageParseError'>\n",
      "        <class 'email.errors.HeaderParseError'>\n",
      "        <class 'email.errors.BoundaryError'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "      <class 'email.errors.CharsetError'>\n",
      "    <class 'uu.Error'>\n",
      "    <class 'http.client.HTTPException'>\n",
      "      <class 'http.client.NotConnected'>\n",
      "      <class 'http.client.InvalidURL'>\n",
      "      <class 'http.client.UnknownProtocol'>\n",
      "      <class 'http.client.UnknownTransferEncoding'>\n",
      "      <class 'http.client.UnimplementedFileMode'>\n",
      "      <class 'http.client.IncompleteRead'>\n",
      "      <class 'http.client.ImproperConnectionState'>\n",
      "        <class 'http.client.CannotSendRequest'>\n",
      "        <class 'http.client.CannotSendHeader'>\n",
      "        <class 'http.client.ResponseNotReady'>\n",
      "      <class 'http.client.BadStatusLine'>\n",
      "        <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'http.client.LineTooLong'>\n",
      "    <class 'IPython.core.magics.code.InteractivelyDefined'>\n",
      "    <class 'IPython.terminal.embed.KillEmbedded'>\n",
      "    <class 'psutil.Error'>\n",
      "      <class 'psutil.NoSuchProcess'>\n",
      "        <class 'psutil.ZombieProcess'>\n",
      "      <class 'psutil.AccessDenied'>\n",
      "      <class 'psutil.TimeoutExpired'>\n",
      "    <class 'psutil._pslinux._Ipv6UnsupportedError'>\n",
      "    <class 'tornado.queues.QueueEmpty'>\n",
      "    <class 'tornado.queues.QueueFull'>\n",
      "    <class '_pydev_bundle._pydev_saved_modules.DebuggerInitializationError'>\n",
      "    <class 'xml.parsers.expat.ExpatError'>\n",
      "    <class 'xmlrpc.client.Error'>\n",
      "      <class 'xmlrpc.client.ProtocolError'>\n",
      "      <class 'xmlrpc.client.ResponseError'>\n",
      "      <class 'xmlrpc.client.Fault'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.ParseBaseException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseFatalException'>\n",
      "        <class 'pkg_resources._vendor.pyparsing.exceptions.ParseSyntaxException'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.RecursiveGrammarException'>\n",
      "    <class 'pkg_resources.ResolutionError'>\n",
      "      <class 'pkg_resources.VersionConflict'>\n",
      "        <class 'pkg_resources.ContextualVersionConflict'>\n",
      "      <class 'pkg_resources.DistributionNotFound'>\n",
      "      <class 'pkg_resources.UnknownExtra'>\n",
      "    <class '_pydevd_bundle.pydevd_runpy._Error'>\n",
      "    <class '_pydevd_bundle.pydevd_resolver.UnableToResolveVariableException'>\n",
      "    <class '_pydev_bundle.pydev_monkey.InvalidTypeInArgsException'>\n",
      "  <class 'GeneratorExit'>\n",
      "  <class 'SystemExit'>\n",
      "  <class 'KeyboardInterrupt'>\n",
      "  <class 'asyncio.exceptions.CancelledError'>\n",
      "  <class 'pkg_resources._vendor.more_itertools.more.AbortThread'>\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print('  ' * indent + str(exception_class))\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 1)\n",
    "\n",
    "print_exception_hierarchy(BaseException)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dda7466b-7e2b-4624-9db6-93bc84802dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee546cbf-5b72-4542-80e1-d19778e3260d",
   "metadata": {},
   "outputs": [],
   "source": [
    "The ArithmeticError class in Python is the base class for exceptions that arise from arithmetic operations. It serves as a parent class for \n",
    "various specific arithmetic-related exception classes.\n",
    "Two notable exceptions that inherit from ArithmeticError are ZeroDivisionError and OverflowError."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7a8fc7f8-bc18-468c-982d-8320e2333c8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "caught zerodivision error : division by zero\n"
     ]
    }
   ],
   "source": [
    "try :\n",
    "    result= 10/0\n",
    "except ZeroDivisionError as e:\n",
    "        print(f\"caught zerodivision error : {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5bf07fc5-2f84-4f33-b901-f43bce16142b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "try:\n",
    "    huge_number = sys.maxsize ** 10\n",
    "except OverflowError as e:\n",
    "    print(f\"Caught OverflowError: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e59504c-ebac-4312-9129-ec8969df8dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d9350d-761f-4a3e-9ec1-a7c2d877b2a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "The LookupError class in Python is the base class for exceptions that occur when an item is not found in a collection or when an index is out of range.\n",
    "It provides a common parent class for various lookup-related exceptions, making it easier to catch and handle these types of errors in a consistent \n",
    "manner."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4ee51d7b-1c6c-4c3f-8c03-963900f920df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caught KeyError: 'grape'\n"
     ]
    }
   ],
   "source": [
    "my_dict = {\"apple\": 3, \"banana\": 5, \"orange\": 2}\n",
    "\n",
    "try:\n",
    "    count = my_dict[\"grape\"]\n",
    "except KeyError as e:\n",
    "    print(f\"Caught KeyError: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "60496b50-17aa-47c3-b5fd-1b0e074ac631",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caught IndexError: list index out of range\n"
     ]
    }
   ],
   "source": [
    "my_list = [1, 2, 3, 4, 5]\n",
    "\n",
    "try:\n",
    "    value = my_list[10]\n",
    "except IndexError as e:\n",
    "    print(f\"Caught IndexError: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b6f7886-df89-4145-8025-283b37db2bcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "066278b5-2c8a-468c-aab9-e6c90992bf70",
   "metadata": {},
   "outputs": [],
   "source": [
    "ImportError is an exception in Python that is raised when an import statement fails to locate and load the desired module. It typically occurs \n",
    "when Pythoncannot find the specified module in the search paths (sys.path) or when there's an issue with the module's code or dependencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f1e2d6af-6e49-49f3-93bf-d2b99bb8b9fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caught ImportError: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import non_existent_module\n",
    "except ImportError as e:\n",
    "    print(f\"Caught ImportError: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b58eb395-2824-4efc-bfe9-aaf3905c88dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "ModuleNotFoundError:\n",
    "ModuleNotFoundError is a subclass of ImportError that specifically indicates that the module being imported could not be found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "69502399-0460-4e87-a573-60ccdd3c696d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caught ModuleNotFoundError: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import non_existent_module\n",
    "except ModuleNotFoundError as e:\n",
    "    print(f\"Caught ModuleNotFoundError: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7042ef0-4332-42c3-a9ce-7d4082baae4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "684d8201-3375-4eb2-8da5-f80fd9a9d3db",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dce97c9-a222-47fc-af5d-52fc812570a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
